juju/juju-controller
# Add Loki support

This PR adds a `logging` relation so that the controller charm can send logs to `loki-k8s` (part of the COS Lite stack).

## QA steps

Assuming Juju 3.0 with LXD and MicroK8s set up.

Deploy COS Lite on MicroK8s:
```console
$ juju bootstrap microk8s k8s
$ juju add-model cos
$ juju deploy cos-lite --channel=edge --trust
$ juju offer loki:logging
Application "loki" endpoints [logging] available at "admin/cos.loki"
```

Bootstrap a LXD controller with this controller charm:
```console
$ juju bootstrap lxd lxd --controller-charm-path=[path/to/cloned/repo]
$ juju switch controller
$ juju relate controller k8s:cos.loki
```

Now the Juju controller should be connected to Loki. We will verify by using the Grafana dashboard.
```console
$ juju switch k8s
$ juju status
Unit             Workload  Agent  Address      Ports  Message         
grafana/0*       active    idle   [ip-address]
$ juju run grafana/0 get-admin-password
admin-password: .....
```

Go to `[ip-address]:3000` in your browser, and log into Grafana with username `admin` and the previously obtained password. On the sidebar, go to "Explore" (the compass), and at the top of the page, change the source to Loki. You should be able to see controller logs using the "log browser".

## Notes / questions
- This has only been tested for LXD controllers, it probably won't work for k8s (probably should use Pebble to run Promtail in this case?)
- Need to add proper exception handling in the code.
- Should think about whether the `scrape_configs` setting for Promtail is right - which log files do we want to look at?
